File: quick_sort_step.py
import csv

#  Read a segment of the CSV file specified by the start and end row indices
#  It returns a list of lists containing the data from the specified rows
def read_csv_segment(file_name, start_row, end_row):
    data = []
    with open(file_name, newline='') as csvfile:
        reader = csv.reader(csvfile)
        #  Read the rows in the specified range
        for i, row in enumerate(reader):
            #  If the row is in the specified range, add it to the data list
            if start_row <= i <= end_row:
                #  Convert the row to a list of integers and strings
                number = int(row[0])
                #  Remove any line breaks from the random string
                random_str = row[1]
                data.append([number, random_str])
    return data

def format_data(arr):
    # Format the array as: [num1/str1, num2/str2, ...]
    return "[" + ", ".join(f"{num}/{s}" for num, s in arr) + "]"

def quick_sort_with_steps(arr, low, high, steps):
    if low < high:
        # Print current segment before partitioning
        steps.append(format_data(arr[low:high+1]))

        pi = partition(arr, low, high)
        # Print partition index and array after partitioning
        steps.append(f"pi={pi} {format_data(arr)}")

        quick_sort_with_steps(arr, low, pi - 1, steps)
        quick_sort_with_steps(arr, pi + 1, high, steps)

def partition(arr, low, high):
    #  Choose the last element as the pivot 
    pivot = arr[high][0]
    i = low - 1
    #  Iterate through the array
    for j in range(low, high):
        #  If the current element is less than the pivot, swap it with the element at index i
        if arr[j][0] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    #  Swap the pivot element with the element at index i+1
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def write_steps_to_file(steps, start_row, end_row):
    filename = f"quick_sort_step_{start_row}_{end_row}.txt"
    with open(filename, 'w') as f:
        for step in steps:
            f.write(step + '\n')
    print(f"\n Sorting steps saved to: {filename}")

# --- Main Execution ---    
if __name__ == "__main__":
    try:
        #  Get the start and end row numbers from the user
        start_row = int(input("Enter the START row number (0-based index): "))
        end_row = int(input("Enter the END row number (0-based index): "))

        #  Check if the range is valid
        if start_row < 0 or end_row < start_row:
            raise ValueError("Invalid range. START must be >= 0 and END >= START.")
        
        #  Get the input file name from the user
        input_file = "dataset_5000.csv"

        #  Read the data segment from the CSV file
        data_segment = read_csv_segment(input_file, start_row, end_row)
        if not data_segment:
            raise ValueError("No data found in the given row range.")

        steps = []
        quick_sort_with_steps(data_segment, 0, len(data_segment) - 1, steps)
        write_steps_to_file(steps, start_row, end_row)

    except Exception as e:
        print(f"Error: {e}")

File: quick_sort_step.cpp
#include <iostream>      // for input and output
#include <fstream>       // for file operations     
#include <sstream>       // for string stream operations
#include <vector>        // for vector container
#include <string>        // for string operations
#include <algorithm>     // for std::swap

using namespace std;

// Struct to store each row of data
struct DataRow {
    int number;
    string random_string;
};

// Function to read CSV file from start_row to end_row
// This function reads a segment of the CSV file specified by the start and end row indices
// It returns a vector of dataset structs containing the data from the specified rows

vector<DataRow> read_csv_segment(const string& filename, int start_row, int end_row) {
    vector<DataRow> dataset;  
    ifstream file(filename);
    string line;
    int current_row = 0;   // Initialize current row index

    while (getline(file, line)) {
        if (current_row >= start_row && current_row <= end_row) { //Check if current row is within the specified range
            stringstream ss(line);
            string num_str, rand_str; // Initialize string variables to store number and random string  
            if (getline(ss, num_str, ',') && getline(ss, rand_str)) { //Extract number and random string from the line
                DataRow d;    
                d.number = stoi(num_str);
                // Remove any line breaks from the random string
                rand_str.erase(remove(rand_str.begin(), rand_str.end(), '\n'), rand_str.end());
                rand_str.erase(remove(rand_str.begin(), rand_str.end(), '\r'), rand_str.end());
                d.random_string = rand_str;
                dataset.push_back(d);
            }
        }
        current_row++;
        if (current_row > end_row) break;
    }

    return dataset;
}

// Helper function to format array segment as string
// Display the index from low to high
string format_array(const vector<DataRow>& arr, int low, int high) {
    stringstream ss;
    ss << "[";
    for (int i = low; i <= high; i++) {
        ss << arr[i].number << "/" << arr[i].random_string;
        if (i < high) ss << ", ";
    }
    ss << "]";
    return ss.str();
}

// Partition function where last element is chosen as pivot
int partition(vector<DataRow>& arr, int low, int high) {
    int pivot = arr[high].number;
    int i = low - 1;
    for (int j = low; j < high; j++) {
        // If current element is less than pivot, swap it with the element at index i
        if (arr[j].number < pivot) {
            i++;
            std::swap(arr[i], arr[j]);
        }
    }
    // Swap the pivot element with the element at index i+1
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

// Quick sort function with output steps tracking
void quick_sort(vector<DataRow>& arr, int low, int high, vector<string>& steps) {
    if (low < high) {
        // Format the array segment as a string and add it to the steps vector
        steps.push_back(format_array(arr, low, high));

        int pi = partition(arr, low, high);
        // Add the partition index and the formatted array to the steps vector
        steps.push_back("pi=" + to_string(pi) + " " + format_array(arr, low, high));

        quick_sort(arr, low, pi - 1, steps);
        // Recursively sort the elements before and after the partition index
        quick_sort(arr, pi + 1, high, steps);
    }
}

// Write steps to file
void write_steps_to_file(const vector<string>& steps, int start_row, int end_row) {
    string filename = "quick_sort_step_" + to_string(start_row) + "_" + to_string(end_row) + ".txt";
    ofstream out(filename);
    for (const string& step : steps) {
        out << step << endl;
    }
    cout << "\nSorting steps saved to: " << filename << endl;
}

int main() {
    int start_row, end_row;
    string input_file = "dataset_5000.csv";

    cout << "Enter the START row number (0-based index): ";
    cin >> start_row;
    cout << "Enter the END row number (0-based index): ";
    cin >> end_row;

    if (start_row < 0 || end_row < start_row) {
        cerr << "Error: Invalid row range." << endl;
        return 1;
    }

    vector<DataRow> dataset = read_csv_segment(input_file, start_row, end_row);
    if (dataset.empty()) {
        cerr << "Error: No data found in the given row range." << endl;
        return 1;
    }

    vector<string> steps;
    quick_sort(dataset, 0, dataset.size() - 1, steps);
    write_steps_to_file(steps, start_row, end_row);

    return 0;
}